crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,495 @@
+        ));
+    }
+
+    checker.diagnostics.push(diagnostic);
+}
+
+struct DunderAllValue {
+    items: Vec<DunderAllItem>,
+    range: TextRange,
+    multiline: bool,
+}
+
+impl DunderAllValue {
+    fn from_expr(value: &ast::Expr, locator: &Locator) -> Option<DunderAllValue> {
+        // Step (1): inspect the AST to check that we're looking at something vaguely sane:
+        let is_multiline = locator.contains_line_break(value.range());
+        let (elts, range) = match value {
+            ast::Expr::List(ast::ExprList { elts, range, .. }) => (elts, range),
+            ast::Expr::Tuple(ast::ExprTuple { elts, range, .. }) => (elts, range),
+            _ => return None,
+        };
+        // An `__all__` definition with <2 elements can't be unsorted;
+        // no point in proceeding any further here
+        if elts.len() < 2 {
+            return None;
+        }
+        for elt in elts {
+            // Only consider sorting it if __all__ only has strings in it
+            let string_literal = elt.as_string_literal_expr()?;
+            // And if any strings are implicitly concatenated, don't bother
+            if string_literal.value.is_implicit_concatenated() {
+                return None;
+            }
+        }
+        // Step (2): parse the `__all__` definition using the raw tokens.
+        //
+        // (2a). Start by collecting information on each line individually:
+        let lines = collect_dunder_all_lines(*range, locator)?;
+
+        // (2b). Group lines together into sortable "items":
+        //   - Any "item" contains a single element of the `__all__` list/tuple
+        //   - "Items" are ordered according to the element they contain
+        //   - Assume that any comments on their own line are meant to be grouped
+        //     with the element immediately below them: if the element moves,
+        //     the comments above the element move with it.
+        //   - The same goes for any comments on the same line as an element:
+        //     if the element moves, the comment moves with it.
+        let items = collect_dunder_all_items(&lines);
+
+        Some(DunderAllValue {
+            items,
+            range: *range,
+            multiline: is_multiline,
+        })
+    }
+
+    fn construct_sorted_all(
+        &self,
+        locator: &Locator,
+        parent: &ast::Stmt,
+        stylist: &Stylist,
+    ) -> SortedDunderAll {
+        let mut sorted_items = self.items.clone();
+        sorted_items.sort();
+
+        // As well as saving us unnecessary work,
+        // returning early here also means that we can rely on the invariant
+        // throughout the rest of this function that both `items` and `sorted_items`
+        // have length of at least two. If there are fewer than two items in `__all__`,
+        // it is impossible for them *not* to compare equal here:
+        if sorted_items == self.items {
+            return SortedDunderAll::AlreadySorted;
+        }
```

Comment:
You can use `self.items.is_sorted()` to test if the items are sorted (and avoids cloning all elements). This has the advantage that checking if the items are sorted is `O(n)` instead of `O(n log(n))` (sorting) + `O(n)` (comparing). The complexity remains the same if the items aren't sorted because you have to call `self.items.sort()` to sort the items but that should be less common for projects actively using the lint rule.